docs/tutorials/react-native.rst

```diff
@@ -96,42 +80,84 @@ Let's translate the screen heading:
  <Text style={styles.heading}><Trans>Message Inbox</Trans></Text>
 
 
-This was easy. Now, the next step is to translate the ``title`` prop of the ``Button`` component. But wait a sec, the button expects to receive a ``string``, so we cannot use the ``<Trans>`` component here! Also notice the same holds also for the ``Alert.alert`` call.
+This was easy. Now, the next step is to translate the ``title`` prop of the :component:`Button` component. But wait a sec, the button expects to receive a ``string``, so we cannot use the :component:`Trans` component here! Also notice that the ``Alert.alert`` call requires a string as well.
+
+At this point, we need to turn our attention to the ``@lingui/core`` package, namely the ``setupI18n`` method which gives us an ``i18n`` object that we can use like this: ``i18n.t`this will be translated``` and the result of such call is a string. Let's get to it!
+
+.. hint::
 
-At this point, we need to turn our attention to the ``@lingui/core`` package, namely the ``setupI18n`` method which can give us an object that we can use like this: ``i18n.t`this will be translated``` and the result of such call is a string. Let's get to it!
+  The ``i18n`` object is covered in greater detail in the :ref:`JavaScript tutorial <js-tutorial-label>`.
+
+.. code-block:: jsx
+
+  import { setupI18n } from '@lingui/core';
+
+  // this file is generated by the cli
+  import enMessages from './locale/en/messages.js';
+
+  let i18n = setupI18n({
+   language: 'en',
+   catalogs: {
+     en: enMessages,
+   },
+  });
+  ...
+  <Button onPress={markAsRead} title={i18n.t`mark messages as read`} />
```

Comment:
What I mean here is that you don't need to use `setupI18n` when you're working with translations **inside** your components. You need to work with standalone `i18n` object only outside components, e.g.: in redux-saga actions, etc.

In components, you're better off to wrap you component in `withI18n` decorator and then you injected `i18n` prop:

```js

import { withI18n } from '@lingui/react'

function MarkAsRead({ i18n, markAsRead }) {
  return <Button onPress={markAsRead} title={i18n.t`mark messages as read`} />
}

// withI18n expects options as a first argument,
// so we need to call it with empty parenthesis to get the actual HOC
export default withI18n()(MarkAsRead)
```

Reply:
okay, gotcha. Then my question is *why*?

Also, most apps need both - accessing the i18n in components and outside of them - so this approach, if it is valid (?) need to be documented *somewhere*. I'm okay with using just `this.props.i18n.t` here because we only deal with components, but then we need to do another guide that will cover using both using `i18n.t` and `this.props.i18n.t` and explain **why** using `this.props.i18n.t` in components is better that just `i18n.t` because I'm a lazy person and don't understand why I should use something from props if I can just import a function (`i18n.t`) and use it everywhere. Hope you understand me. Thanks!

Reply:
I do. The main reason is re-rendering. When you directly call a function inside a component, the component itself won't re-render when the return value changes because of outside effects.

```jsx
function TranslatedComponent() {
   return <span>{i18n.t`Hello World`}</span>
}
```

When you render this component, it will show translation for `Hello World`. However, when you change the active language, this component will not re-render. This was common issue in `react-intl` and the only known workaround is to pass `key={language}` to top-most component and force complete re-render of the whole app.

On the other hand, when you use `withI18n` decorator:

```jsx
const TranslatedComponent = withI18n()(({ i18n }) => {
   return <span>{i18n.t`Hello World`}</span>
})
```

Now `withI18n` HOC is responsible not only for injecting `i18n` object, but also for forcing re-render when language or message catalog changes.

Other reason I might thing off: when you use `i18n` directly, you're using global variable, which makes component very hard for testing. I always use functions/variables passed as props instead of global ones.

What do you think? I'm really open for discussion here as [there should be one and preferably only one way to do it](https://zen-of-python.info/there-should-be-one-and-preferably-only-one-obvious-way-to-do-it.html#13).

Reply:
okay, that sounds good. So the recommended approach for projects where both component and non-component code needs to be translated is to 
1 . call `setupI18n` to get the i18n object, 
2 . pass it to i18nProvider and use this.props.i18n in components
3 . use the original i18n object outside of components

?

Reply:
Yes! 

In my project, I don't need i18n outside components. I'm working on React/Apollo app and everything is either a Component (React) or HOC (Apollo containers). AFAIK even redux-form has access to `prop` in validation function (my previous project) so I always considered using standalone i18n object as something special :)

Reply:
Good point! I haven't worked with react-native yet.

I think this section is perfect place to document it. Later we might write a standalone guide to further explain differences and reasoning. For tutorial purposes, let's keep it simple.

Reply:
One more question: we initially need to call `setupI18n` to get the i18n object. Later if I want to change the language, I could create a new i18n object by again calling `setupI18n`, or I can call `i18n.load` and `i18n.activate` like [here](https://github.com/lingui/js-lingui/blob/1f000eb73f0d434885157393420a49ca355a1b93/packages/core/src/i18n.js#L201). Which way should be documented? I kinda like the second one better but not sure if that api is stable..?